### PR TITLE
Add parsing of the PE Authenticode format

### PR DESCRIPTION
**Pull request by HoundThe**
I'm submitting progress on Authenticode parsing, it's still a work in progress.

There are few shortcuts I've taken during development that I need to finish.

- [x] Parse out the rest of the information from PKCS9 counter-signatures
- [x] Parse undocumented MsCounterSignature 
- [x] Check if the signer can have multiple counter signers, same with a counter signer
- [x] Free the Openssl allocations correctly (use smart pointers)
- [x] Implement plain text output
- [x]  Add verification of the signatures and export any abnormalities to the user output

Also, the commit history is a mess, I can fix it if that's a problem.

Further plans after this could be adding more things that are being parsed, validation/verification of the signature contents

Also, solves #380.

### CODE REVIEW

File: src/fileformat/file_format/pe/authenticode/authenticode_structs.h [hunk 1]
```diff
@@ -0,0 +1,91 @@
+/* Author Marek Milkovič - metthal */
```
Comment (PeterMatula) [inline src/fileformat/file_format/pe/authenticode/authenticode_structs.h:1]: Again, proper header. If you want to keep the author, you can do so in the Doxygen format. This is not standard in the other sources, but in this case @metthal can be mentioned.

File: src/fileformat/file_format/pe/authenticode/helper.cpp [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+		return nullptr;
+	}
+
+	PKCS7* pkcs7 = d2i_PKCS7_bio(bio, nullptr);
+
+	BIO_free(bio);
+	return pkcs7;
+}
+
+} // namespace authenticode
```
Comment (PeterMatula) [inline src/fileformat/file_format/pe/authenticode/helper.cpp:114]: new line

File: src/fileformat/file_format/pe/authenticode/helper.h [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+std::string bytesToHexString(const std::uint8_t* in, int len);
+std::string parsePublicKey(BIO* bio);
+std::string serialToString(ASN1_INTEGER* serial);
+std::string X509NameToString(X509_NAME* name);
+std::string parseDateTime(const ASN1_TIME* dateTime);
+PKCS7* getPkcs7(const std::vector<unsigned char>& input);
+void calculateDigest(const EVP_MD* md, const std::uint8_t* data, int len, std::uint8_t* digest);
+
+} // namespace authenticode
```
Comment (PeterMatula) [inline src/fileformat/file_format/pe/authenticode/helper.h:30]: new line

File: src/fileformat/file_format/pe/authenticode/ms_counter_signature.cpp [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+	is_valid = PKCS7_signatureVerify(p7bio, pkcs7.get(), si, const_cast<X509*>(signCert)) == 1;
+	if (!is_valid) {
+		warnings.emplace_back("Failed to verify the counter-signature");
+	}
+
+	BIO_free_all(p7bio);
+	return warnings;
+}
+
+} // namespace authenticode
```
Comment (PeterMatula) [inline src/fileformat/file_format/pe/authenticode/ms_counter_signature.cpp:121]: new line, got through your sources and add it everywhere when missing

File: src/fileformat/file_format/pe/authenticode/x509_certificate.cpp [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+	out_cert.sha256Digest = getSha256();
+	out_cert.validSince = getValidSince();
+	out_cert.validUntil = getValidUntil();
+	return out_cert;
+}
+
+// /* It's assumed that the certificates are processed in correct
+//    order by callbacks set in the CertProcessor constructor */
+
+// static CertificateProcessor* getProcessor(X509_STORE_CTX* ctx)
```
Comment (PeterMatula) [inline src/fileformat/file_format/pe/authenticode/x509_certificate.cpp:217]: we generally don't want to have this kind of dead code in the source base. Is this helpful in any way? Could it be removed?

File: src/fileformat/file_format/pe/pe_format.cpp [hunk 6]
```diff
@@ -612,12 +576,12 @@ bool verifySignature(const retdec::fileformat::PeFormat* peFile, PKCS7 *p7)
 	}
 
 	auto storedHash = std::static_pointer_cast<Asn1OctetString>(digestValue)->getString();
-	auto calculatedHash = calculateDigest(peFile, algorithm);
-	if (storedHash != calculatedHash)
-	{
-		EVP_cleanup();
-		return false;
-	}
+	// auto calculatedHash = calculateDigest(peFile, algorithm);
```
Comment (PeterMatula) [inline src/fileformat/file_format/pe/pe_format.cpp:579]: dead code ...

File: src/fileinfo/file_information/file_information.cpp [hunk 7]
```diff
@@ -1628,460 +1655,452 @@ bool FileInformation::hasResourceTableRecords() const
...
-
-/**
- * Find out if there is counter-signer certificate
- * @return @c true if there is counter-signer certificate, @c false otherwise
- */
-bool FileInformation::hasCertificateTableCounterSignerCertificate() const
-{
-	return certificateTable.hasCounterSignerCertificate();
-}
+// /**
```
Comment (PeterMatula) [inline src/fileinfo/file_information/file_information.cpp:1659]: So much dead code!
Was it moved somewhere? Is it not needed anymore?
Unless it is super useful to have this here, remove it.

File: src/fileinfo/file_information/file_information.h [hunk 8]
```diff
@@ -23,637 +23,595 @@ namespace fileinfo {
...
-		void removeRedundantCryptoRules();
-		void sortCryptoPatternMatches();
-		void addMalwarePattern(Pattern &pattern);
-		void sortMalwarePatternMatches();
-		void addOtherPattern(Pattern &pattern);
-		void sortOtherPatternMatches();
-		void addTool(retdec::cpdetect::DetectResult &tool);
-		void addLoadedSegment(const LoadedSegment& segment);
-		/// @}
+private:
```
Comment (PeterMatula) [inline src/fileinfo/file_information/file_information.h:26]: It is unfortunate to include this kind of change in a PR. It is needlessly convoluting things.

**Comment by HoundThe** [comment]
Yes, I polluted the PR when I merged with master, I removed the polluting commits and forced pushed it into (hopefully) correct state.

**Comment by PeterMatula** [comment]
Now the tests are falling, probably only because of `"certificateTable" -> `"digitalSignatures"` rename. I hope once https://github.com/avast/retdec-regression-tests/pull/86 is modified and notes there are fixed it will pass and we can merge it.

**Comment by PeterMatula** [comment]
Lets run TC tests.

**Comment by PeterMatula** [comment]
TC tests are still failing, but it is because it is not testing this PR with the [associated tests PR](https://github.com/avast/retdec-regression-tests/pull/86). For future reference - RetDec PR branch is tested with the RetDec regression tests PR branch with a matching (i.e. the same) name - if it exists, otherwise with `master`. So in this case `HoundThe:sig_parser` is not matching `HoundThe:authenticode`.

**Comment by PeterMatula** [comment]
@HoundThe can you try to rename the branch for [regression tests PR](https://github.com/avast/retdec-regression-tests/pull/86) to `HoundThe:sig_parser` and create a new PR? It should work if there won't be another problem with the fact that these are branches from another repo (your clone). We use PRs even for merging our branches from this repo and I'm not sure we ever tested it with 3rd party PR from cloned repos.

**Comment by HoundThe** [comment]
I've renamed the PR branch to `sig_parser`

**Comment by HoundThe** [comment]
Some Windows tests were failing because of UB from an uninitialized algorithm value, when parsing of a countersignature fails, should be fixed now.

**Pull request merged**